server/src/main/java/org/elasticsearch/action/support/single/shard/TransportSingleShardAction.java

```diff
@@ -175,24 +192,31 @@ public void start() {
                     new ActionListenerResponseHandler<>(listener, reader)
                 );
             } else {
-                perform(null);
+                perform(null, null);
             }
         }
 
-        private void onFailure(ShardRouting shardRouting, Exception e) {
+        private void tryNextShardCopy(ShardRouting lastShardRouting, Exception e, Response r) {
             if (e != null) {
-                logger.trace(() -> format("%s: failed to execute [%s]", shardRouting, internalRequest.request()), e);
+                logger.trace(() -> format("%s: failed to execute [%s]", lastShardRouting, internalRequest.request()), e);
             }
-            perform(e);
+            perform(e, r);
         }
 
-        private void perform(@Nullable final Exception currentFailure) {
+        private void perform(@Nullable final Exception currentFailure, @Nullable final Response currentResponse) {
             Exception lastFailure = this.lastFailure;
             if (lastFailure == null || TransportActions.isReadOverrideException(currentFailure)) {
                 lastFailure = currentFailure;
                 this.lastFailure = currentFailure;
             }
-            final ShardRouting shardRouting = shardIt.nextOrNull();
+            this.lastResponse = currentResponse;
+
+            ShardRouting shardRouting = shardIt.nextOrNull();
+            if (shardRouting != null) {
+                while (shardRouting != null && skipShardCopy(shardRouting, lastResponse, lastFailure)) {
```

Comment:
NIT: it seems this `if` is redundant

Reply:
I guess it is! But I think most of this is going to change anyway!